sdk/how-to-guides/ports.md

```diff
@@ -0,0 +1,144 @@
+
+Now, in the body of the charm definition, define the event handler, and set the port that should be open - this defaults to a TCP port, but UDP and ICMP can also be specified. For example:
+
+```python
+def _on_config_changed(self, event: ops.ConfigChangedEvent):
+    self.unit.set_ports(self.config["server-port"])
+```
+
+> Examples: [`loki-k8s` sets the open port in the charm `__init__`](https://github.com/canonical/loki-k8s-operator/blob/2e4674d64b692f82ae3dec8f7f1f0745f087827f/src/charm.py#L150), [`mysql-router` sets ports based on whether the charm is configured to be externally accessible](https://github.com/canonical/mysql-router-operator/blob/04c7b3a0ebb6279648bd3f72c5c04b0f7bad1e0c/src/machine_charm.py#L111)
+
+> See more: [`ops.Unit.set_ports`](https://ops.readthedocs.io/en/latest/#ops.Unit.set_ports)
+
+<a href="#heading--test-the-port-is-opened"><h2 id="heading--test-the-port-is-opened">Test the port is opened</h2></a>
+
+> See first: [Get started with charm testing](https://juju.is/docs/sdk/get-started-with-charm-testing)
+
+You'll want to add three levels of tests, unit, Scenario, and integration.
+
+- [Test setting the workload version -- unit tests](#heading--write-unit-tests)
+- [Test setting the workload version -- Scenario tests](#heading--write-scenario-tests)
+- [Test setting the workload version -- integration tests](#heading--write-integration-tests)
+
+<a href="#heading--write-unit-tests"><h3 id="heading--write-unit-tests">Write unit tests</h3></a>
+
+> See first: [How to write unit tests for a charm](/t/4461)
+
+To use a unit test to verify that updating the charm config opens a port, for this charm that opens a port based on the config, the test needs to trigger the `config-changed` event and then check which ports are open. In your `tests/unit/test_charm.py` file, add the following test function to the file:
+
+```python
+def test_port_configuration():
+    harness = ops.testing.Harness()
+    harness.begin()
+
+    port = 8080
+    harness.update_config({"server-port": port})
+
+    assert harness.model.unit.opened_ports() == {ops.Port("tcp", port)}
+```
+
+> Examples: [`charm-microk8s` checks that the `install` hook opens a port](https://github.com/canonical/charm-microk8s/blob/1e7fd26e4f85f361a813106325d84bf079cd7b1a/tests/unit/test_charm_control_plane.py#L34)
+
+> See more: [`ops.Unit.opened_ports`](https://ops.readthedocs.io/en/latest/#ops.Unit.opened_ports)
+
+<a href="#heading--write-scenario-tests"><h3 id="heading--write-scenario-tests">Write scenario tests</h3></a>
+
+> See first: [How to write scenario tests for a charm](/t/10585)
+
+To use a Scenario test to verify that the `config-changed` event results in the port being opened, the new config is passed to the `State`, and the `State.opened_ports` attribute is checked after running the event. For example, in your `tests/scenario/test_charm.py` file, add the following test function to the file:
```

Comment:
Rephrase suggestion:


(1) 
> pass the new config to the `State` and, after running the event, check the `State.opened_ports` attribute

Active voice tends to be both clearer and faster to parse. 

(2) 
> add the following test function:

"to the file" is redundant.